contrib/epee/include/net/abstract_tcp_server2.inl

```diff
@@ -81,860 +81,1019 @@ namespace net_utils
-    // handler returns. The connection class's destructor closes the socket.
-    CATCH_ENTRY_L0("connection<t_protocol_handler>::handle_receive", void());
+  template<typename T>
+  void connection<T>::on_terminating()
+  {
+    assert(state.status == status_t::TERMINATING);
+    if (state.timers.general.wait_expire)
+      return;
+    if (state.socket.wait_handshake)
+      return;
+    if (state.timers.throttle.in.wait_expire)
+      return;
+    if (state.socket.wait_read)
+      return;
+    if (state.socket.handle_read)
+      return;
+    if (state.timers.throttle.out.wait_expire)
+      return;
+    if (state.socket.wait_write)
+      return;
+    if (state.socket.wait_shutdown)
+      return;
+    if (state.protocol.wait_init)
+      return;
+    if (state.protocol.wait_callback)
+      return;
+    if (state.protocol.wait_release)
+      return;
+    if (state.socket.connected) {
+      ec_t ec;
+      connection_basic::socket_.next_layer().shutdown(
+        socket_t::shutdown_both,
+        ec
+      );
+      connection_basic::socket_.next_layer().close(ec);
+      state.socket.connected = false;
+    }
+    state.status = status_t::WASTED;
   }
-  //---------------------------------------------------------------------------------
-  template<class t_protocol_handler>
-  bool connection<t_protocol_handler>::call_run_once_service_io()
+
+  template<typename T>
+  bool connection<T>::send(byte_slice_t message)
   {
-    TRY_ENTRY();
-    if(!m_is_multithreaded)
-    {
-      //single thread model, we can wait in blocked call
-      size_t cnt = GET_IO_SERVICE(socket()).run_one();
-      if(!cnt)//service is going to quit
-        return false;
-    }else
-    {
-      //multi thread model, we can't(!) wait in blocked call
-      //so we make non blocking call and releasing CPU by calling sleep(0); 
-      //if no handlers were called
-      //TODO: Maybe we need to have have critical section + event + callback to upper protocol to
-      //ask it inside(!) critical region if we still able to go in event wait...
-      size_t cnt = GET_IO_SERVICE(socket()).poll_one();
-      if(!cnt)
-        misc_utils::sleep_no_w(1);
+    lock_guard_t guard(state.lock);
+    if (state.status != status_t::RUNNING || state.socket.wait_handshake)
+      return false;
+    auto random_delay = []{
+      using engine = std::mt19937;
+      std::random_device dev;
+      std::seed_seq::result_type rand[
+        engine::state_size // Use complete bit space
+      ]{};
+      std::generate_n(rand, engine::state_size, std::ref(dev));
+      std::seed_seq seed(rand, rand + engine::state_size);
+      engine rng(seed);
+      return std::chrono::milliseconds(
+        std::uniform_int_distribution<>(5000, 6000)(rng)
+      );
+    };
+    if (state.data.write.wait_consume ||
+      state.data.write.queue.size() > ABSTRACT_SERVER_SEND_QUE_MAX_COUNT
+    )
+      state.condition.wait_for(state.lock, random_delay(),
```

Comment:
This `wait_for` is going to need some justification, in the comments, and with some performance testing. Or just any explanation at all. Because this is slowing down all outgoing writes by an arbitrary amount of time.

Reply:
Wait this was inside an `if` statement? Ugh, sorry, can we fix the style of this at least?

Reply:
Ok I see why this was duplicated, but there has to be a simpler more concise technique. And seriously `{}` this!

Reply:
Updated